libcudacxx/docs/extended_api/asynchronous_operations/memcpy_async_tx.md

```diff
@@ -0,0 +1,100 @@
+---
+grand_parent: Extended API
+parent: Asynchronous Operations
+---
+
+# `cuda::device::memcpy_async_tx`
+
+Defined in header `<cuda/barrier>`:
+
+```cuda
+template <typename T, size_t Alignment>
+inline __device__
+async_contract_fulfillment
+cuda::device::memcpy_async_tx(
+  T* dest,
+  const T* src,
+  cuda::aligned_size_t<Alignment> size,
+  cuda::barrier<cuda::thread_scope_block>& bar);
+```
+
+Copies `size` bytes from global memory `src` to shared memory `dest` and decrements the transaction count of `bar` by `size` bytes.
+
+## Preconditions
+
+* `src`, `dest` are 16-byte aligned and `size` is a multiple of 16, i.e.,
+  `Alignment >= 16`.
+* `dest` points to a shared memory allocation that is at least `size` bytes wide.
+* `src` points to a global memory allocation that is at least `size` bytes wide.
+* `bar` is located in shared memory
+* If either `destination` or `source` is an invalid or null pointer, the
+    behavior is undefined (even if `count` is zero).
+
+## Requires
+
+* `is_trivially_copyable_v<T>` is true.
+ 
+## Notes
+
+This function can only be used under CUDA Compute Capability 9.0 (Hopper) or
+higher.
+
+There is no feature flag to check if `cuda::device::memcpy_async_tx` is
+available.
+
+**Comparison to `cuda::memcpy_async`**: `memcpy_async_tx` supports a subset of
+the operations of `memcpy_async`. It gives more control over the synchronization
+with a barrier than `memcpy_async`. Currently, `memcpy_async_tx` has no synchronous
+fallback mechanism., i.e., it currently does not work on older hardware
+(pre-CUDA Compute Capability 9.0, i.e., Hopper).
+
+## Return Value
+
+Returns `async_contract_fulfillment::async`.
+
+## Example
+
+```cuda
+#include <cuda/barrier>
+#include <cuda/std/utility> // cuda::std::move
+
+#if defined(__CUDA_MINIMUM_ARCH__) && __CUDA_MINIMUM_ARCH__ < 900
+static_assert(false, "Insufficient CUDA Compute Capability: cuda::device::memcpy_async_tx is not available.");
+#endif // __CUDA_MINIMUM_ARCH__
+
+__device__ alignas(16) int gmem_x[2048];
+
+__global__ void example_kernel() {
+  __shared__ alignas(16) int smem_x[1024];
+  __shared__ cuda::barrier<cuda::thread_scope_block> bar;
```

Comment:
Ditto adapt the example to create a painter and init that

Reply:
This is documentation though? cuda_space_selector is not available here?

Reply:
🤦 yeah just leave it as is